src/wbase/ChannelShared.h

```diff
@@ -0,0 +1,248 @@
+                                        util::MultiError& multiErr, std::atomic<bool>& cancelled) = 0;
+
+    /// @return a log worthy string describing transmitData.
+    std::string dumpTransmit() const;
+
+protected:
+    /// Protected constructor is seen by subclasses only.
+    ChannelShared(std::shared_ptr<wbase::SendChannel> const& sendChannel,
+                  std::shared_ptr<wcontrol::TransmitMgr> const& transmitMgr, qmeta::CzarId czarId);
+
+    std::shared_ptr<wbase::SendChannel> const sendChannel() const { return _sendChannel; }
+
+    /// Dumps transmitData into a string within the thread-safe context.
+    /// @param tMtxLock - Lock on mutex tMtx to be acquired before calling the method.
+    std::string dumpTransmit(std::lock_guard<std::mutex> const& tMtxLock) const;
+
+    /// @return a new TransmitData::Ptr object.
+    /// @param tMtxLock - Lock on mutex tMtx to be acquired before calling the method.
+    TransmitData::Ptr createTransmit(std::lock_guard<std::mutex> const& tMtxLock, Task& task);
+
+    /// Create a new transmitData object if needed.
+    /// @param tMtxLock - Lock on mutex tMtx to be acquired before calling the method.
+    void initTransmit(std::lock_guard<std::mutex> const& tMtxLock, Task& task);
+
+    /// Prepare the transmit data and then call addTransmit.
+    /// @param tMtxLock - Lock on mutex tMtx to be acquired before calling the method.
+    bool prepTransmit(std::lock_guard<std::mutex> const& tMtxLock, std::shared_ptr<Task> const& task,
+                      bool cancelled, bool lastIn);
+
+    /// Try to transmit the data in tData.
+    /// If the queue already has at least 2 TransmitData objects, addTransmit
+    /// may wait before returning. Result rows are read from the
+    /// database until there are no more rows or the buffer is
+    /// sufficiently full. addTransmit waits until that buffer has been
+    /// sent to the czar before reading more rows. Without the wait,
+    /// the worker may read in too many result rows, run out of memory,
+    /// and crash.
+    /// @param tMtxLock - Lock on mutex tMtx to be acquired before calling the method.
+    /// @return true if transmit was added successfully.
+    /// @see ChannelShared::_transmit code for further explanation.
+    bool addTransmit(std::lock_guard<std::mutex> const& tMtxLock, std::shared_ptr<Task> const& task,
+                     bool cancelled, bool erred, bool lastIn, TransmitData::Ptr const& tData, int qId,
+                     int jId);
+
+    /// Items to share one TransmitLock across all Task's using this
+    /// ChannelShared. If all Task's using this channel are not
+    /// allowed to complete, deadlock is likely.
+    void waitTransmitLock(bool interactive, QueryId const& qId);
+
+    std::shared_ptr<TransmitData> transmitData;  ///< TransmitData object
+    mutable std::mutex tMtx;                     ///< protects transmitData
```

Comment:
Each has its own pros and cons, depending on the context where such patterns are used, In this particular case, it's better to send a reference to the `lock_guard` around since all uses of the corresponding mutexes are confined within the implementation of the class. The lock-enforcement mechanism implemented in the code works at the compile time, and as such, it doesn't require any additional checks to be performed at run-time.
The dynamic (run-time) reinforcement (`util::VMutex `) might be required in other contexts where more complex synchronizations might be needed. Personally, I would really avoid having any code where run-time checking is needed because such code is more difficult to test and debug.